File: DailyCodingProblems/400_sublist_sum_Goldman_Sachs/python/main.py
#coding:utf-8
class Main:
    table = []
    # 初始化Main对象
    def __init__(self,size):
        self.table = [0] * size
    # 获得最低有效位
    def lowbit(self,v):
        return v & (-v)
    # 更新指定下标的值
    def update(self,i,size,val):
        i+=1
        while i <= size:
            self.table[i-1] += val
            i += self.lowbit(i)
    # 获取0~idx区间的和
    def getSum(self,idx):
        idx+=1
        s = 0
        while idx > 0:
            s += self.table[idx-1]
            idx -= self.lowbit(idx)
        return s
    # 获取[i,j)区间的和
    def sum(self,i,j):
        if i >= j or j < 1: # 给定的下标需要合法
            return 0
        return self.getSum(j-1) - self.getSum(i-1)

datas = [1,2,3,4,5]
size = len(datas)
solution = Main(size)
for i in range(size):
    solution.update(i,size,datas[i])
print(solution.sum(0,3))

File: DailyCodingProblems/400_sublist_sum_Goldman_Sachs/c/main.c
#include<stdio.h>

#define SIZE 10000
int table[SIZE];

int lowbit(int num);
void update(int i,int size,int val);
int getSum(int i);
int sum(int i,int j);

int main(){
    int arr[] = {
        1,2,3,4,5,6
    };
    int size = sizeof(arr) / sizeof(arr[0]);
    int i;
    for(i = 0; i < size; i++)
        update(i,size,arr[i]);

    printf("%d\n",sum(1,3));

    return 0;
}

int lowbit(int val){
    return val & (-val);
}

void update(int i,int size,int val){
    i+=1;
    while(i <= size){
        table[i-1] += val;
        i += lowbit(i);
    }
}

int getSum(int i){
    i+=1;
    int s = 0;
    while(i > 0){
        s+=table[i-1];
        i-=lowbit(i);
    }
    return s;
}

int sum(int i,int j){
    return getSum(j-1)-getSum(i-1);
}
